File: KeyBastion-backend/src/main/java/com/jvp/KeyBastion_backend/services/AdminService.java
package com.jvp.KeyBastion_backend.services;

import com.jvp.KeyBastion_backend.dto.AdminDTO.AdminUserCreationRequest;
import com.jvp.KeyBastion_backend.dto.AdminDTO.CategoryStatsResponse;
import com.jvp.KeyBastion_backend.dto.AdminDTO.SharingStatsResponse;
import com.jvp.KeyBastion_backend.model.ActivityLog;
import com.jvp.KeyBastion_backend.model.Category;
import com.jvp.KeyBastion_backend.model.User;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * AdminService defines administrative operations for the KeyBastion application.
 * <p>
 * This service provides methods for managing admin users, viewing activity logs,
 * gathering statistics, and managing global categories. All methods are intended
 * for use by admin users only.
 */
public interface AdminService {
    /**
     * Creates a new admin user in the system.
     *
     * @param request   Details for the new admin user (username, password, etc.)
     * @param ipAddress IP address from which the request was made
     * @return The created admin User object
     */
    User createAdminUser(AdminUserCreationRequest request, String ipAddress);

    /**
     * Retrieves activity logs filtered by action type and date range.
     *
     * @param action    The type of action to filter (e.g., CREATE, UPDATE, DELETE)
     * @param startDate Start of the date range
     * @param endDate   End of the date range
     * @return List of matching ActivityLog entries
     */
    List<ActivityLog> getAllActivityLogs(String action, LocalDateTime startDate, LocalDateTime endDate);

    /**
     * Retrieves all activity logs for a specific user.
     *
     * @param userId The ID of the user
     * @return List of ActivityLog entries for the user
     */
    List<ActivityLog> getUserActivityLogs(UUID userId);

    /**
     * Gets statistics about category usage across all users.
     *
     * @return List of category statistics
     */
    List<CategoryStatsResponse> getCategoryStats();

    /**
     * Gets statistics about credential sharing across the platform.
     *
     * @return List of sharing statistics
     */
    List<SharingStatsResponse> getSharingStats();

    /**
     * Retrieves all global categories available in the system.
     *
     * @return List of global Category objects
     */
    List<Category> getAllCategories();

    /**
     * Creates a new global category that is available to all users.
     *
     * @param category  The category to create
     * @param admin     The admin performing the action
     * @param ipAddress IP address from which the request was made
     * @return The created Category object
     */
    Category createGlobalCategory(Category category, User admin, String ipAddress);

    /**
     * Retrieves a list of all users in the system (for admin view).
     *
     * @return List of user details as key-value maps
     */
    List<Map<String, Object>> getAllUsers();
}

File: KeyBastion-frontend/src/services/adminService.ts
import api from './api';
import { API_ENDPOINTS } from './config';

// Types
export interface AdminUserCreationRequest {
  username: string;
  email: string;
  password: string;
  role: string;
}

export interface CategoryStatsResponse {
  categoryName: string;
  passwordCount: number;
}

export interface SharingStatsResponse {
  categoryName: string;
  shareCount: number;
}

export interface User {
  id: string;
  username: string;
  email: string;
  role: string;
  createdAt: string;
}

export interface ActivityLog {
  id: string;
  userId: string;
  username: string;
  action: string;
  details: string;
  timestamp: string;
  ipAddress?: string;
}

export interface Category {
  id: string;
  name: string;
  description?: string;
  global: boolean;
  userId: string;
}

class AdminService {
  /**
   * Create a new admin user
   */
  public async createAdminUser(userData: AdminUserCreationRequest): Promise<User> {
    try {
      console.log('Creating user with data:', { ...userData, password: '[REDACTED]' });
      
      // Ensure role is uppercase as expected by the backend
      const formattedData = {
        ...userData,
        role: userData.role.toUpperCase()
      };
      
      const response = await api.post<User>(
        API_ENDPOINTS.ADMIN.USERS,
        formattedData,
        { requireAuth: true }
      );
      
      console.log('User creation response:', response);
      return response;
    } catch (error) {
      console.error('Failed to create admin user:', error);
      throw error;
    }
  }

  /**
   * Get all users
   */
  public async getAllUsers(): Promise<User[]> {
    try {
      const response = await api.get<User[]>(
        API_ENDPOINTS.ADMIN.USERS,
        { requireAuth: true }
      );
      return response;
    } catch (error) {
      console.error('Failed to get users:', error);
      throw error;
    }
  }

  /**
   * Get all activity logs
   */
  public async getAllActivityLogs(
    action?: string,
    startDate?: Date,
    endDate?: Date
  ): Promise<ActivityLog[]> {
    try {
      let url = API_ENDPOINTS.ADMIN.ACTIVITY_LOGS;
      const params: string[] = [];
      
      if (action) {
        params.push(`action=${encodeURIComponent(action)}`);
      }
      
      if (startDate) {
        params.push(`startDate=${encodeURIComponent(startDate.toISOString())}`);
      }
      
      if (endDate) {
        params.push(`endDate=${encodeURIComponent(endDate.toISOString())}`);
      }
      
      if (params.length > 0) {
        url = `${url}?${params.join('&')}`;
      }
      
      const response = await api.get<ActivityLog[]>(
        url,
        { requireAuth: true }
      );
      return response;
    } catch (error) {
      console.error('Failed to get activity logs:', error);
      throw error;
    }
  }

  /**
   * Get activity logs for a specific user
   */
  public async getUserActivityLogs(userId: string): Promise<ActivityLog[]> {
    try {
      const response = await api.get<ActivityLog[]>(
        `${API_ENDPOINTS.ADMIN.ACTIVITY_LOGS}/${userId}`,
        { requireAuth: true }
      );
      return response;
    } catch (error) {
      console.error(`Failed to get activity logs for user ${userId}:`, error);
      throw error;
    }
  }

  /**
   * Get category statistics
   */
  public async getCategoryStats(): Promise<CategoryStatsResponse[]> {
    try {
      const response = await api.get<CategoryStatsResponse[]>(
        API_ENDPOINTS.ADMIN.STATS_CATEGORY,
        { requireAuth: true }
      );
      return response;
    } catch (error) {
      console.error('Failed to get category stats:', error);
      throw error;
    }
  }

  /**
   * Get sharing statistics
   */
  public async getSharingStats(): Promise<SharingStatsResponse[]> {
    try {
      const response = await api.get<SharingStatsResponse[]>(
        API_ENDPOINTS.ADMIN.STATS_SHARING,
        { requireAuth: true }
      );
      return response;
    } catch (error) {
      console.error('Failed to get sharing stats:', error);
      throw error;
    }
  }

  /**
   * Get all categories
   */
  public async getAllCategories(): Promise<Category[]> {
    try {
      const response = await api.get<Category[]>(
        API_ENDPOINTS.ADMIN.CATEGORIES,
        { requireAuth: true }
      );
      return response;
    } catch (error) {
      console.error('Failed to fetch categories:', error);
      throw error;
    }
  }

  /**
   * Create a new global category
   */
  public async createGlobalCategory(category: { name: string; description?: string }): Promise<Category> {
    try {
      const response = await api.post<Category>(
        API_ENDPOINTS.ADMIN.CATEGORIES,
        { ...category, global: true },
        { requireAuth: true }
      );
      return response;
    } catch (error) {
      console.error('Failed to create global category:', error);
      throw error;
    }
  }
}

export const adminService = new AdminService();
